be/src/exec/pipeline/pipeline_driver_queue.cpp

```diff
@@ -273,186 +174,233 @@ DriverRawPtr SubQuerySharedDriverQueue::take() {
-                wg->update_last_real_runtime_ns(wg->real_runtime_ns());
-
-                int64_t diff_real_runtime_ns = wg->real_runtime_ns() - origin_real_runtime_ns;
-                workgroup::WorkGroupManager::instance()->increment_cpu_runtime_ns(diff_real_runtime_ns);
-                wg->incr_total_cpu_cost(diff_real_runtime_ns);
+                int64_t new_vruntime_ns = std::min(min_wg_entity->vruntime_ns() - _ideal_runtime_ns(wg_entity) / 2,
+                                                   min_wg_entity->runtime_ns() / int64_t(wg_entity->cpu_limit()));
+                int64_t diff_vruntime_ns = new_vruntime_ns - wg_entity->vruntime_ns();
+                if (diff_vruntime_ns > 0) {
+                    DCHECK(_ready_wg_entities.find(wg_entity) == _ready_wg_entities.end());
+                    wg_entity->incr_runtime_ns(diff_vruntime_ns * wg_entity->cpu_limit());
+                }
             }
         }
-        _ready_wgs.emplace(wg);
+
+        _ready_wg_entities.emplace(wg_entity);
+        _update_min_wg();
     }
-    wg->driver_queue()->put_back(driver);
+
     _cv.notify_one();
 }
 
-workgroup::WorkGroup* DriverQueueWithWorkGroup::_find_min_owner_wg(int worker_id) {
-    workgroup::WorkGroup* min_wg = nullptr;
-    int64_t min_vruntime_ns = 0;
-
-    auto owner_wgs = workgroup::WorkGroupManager::instance()->get_owners_of_driver_worker(worker_id);
-    if (owner_wgs != nullptr) {
-        for (const auto& wg : *owner_wgs) {
-            if (_ready_wgs.find(wg.get()) != _ready_wgs.end() &&
-                (min_wg == nullptr || min_vruntime_ns > wg->vruntime_ns())) {
-                min_wg = wg.get();
-                min_vruntime_ns = wg->vruntime_ns();
-            }
+void WorkGroupDriverQueue::_update_min_wg() {
+    if (_ready_wg_entities.empty()) {
+        _min_vruntime_ns = std::numeric_limits<int64_t>::max();
+        _min_wg_entity = nullptr;
+    } else {
+        auto* min_wg_entity = *_ready_wg_entities.begin();
+        _min_vruntime_ns = min_wg_entity->vruntime_ns();
+        _min_wg_entity = min_wg_entity;
+    }
+}
+
+workgroup::WorkGroupDriverSchedEntity* WorkGroupDriverQueue::_take_next_wg() {
+    workgroup::WorkGroupDriverSchedEntity* min_unthrottled_wg_entity = nullptr;
+    for (auto* wg_entity : _ready_wg_entities) {
+        if (!_throttled(wg_entity)) {
+            min_unthrottled_wg_entity = wg_entity;
+            break;
         }
     }
 
-    return min_wg;
+    return min_unthrottled_wg_entity;
 }
 
-workgroup::WorkGroup* DriverQueueWithWorkGroup::_find_min_wg() {
-    workgroup::WorkGroup* min_wg = nullptr;
-    int64_t min_vruntime_ns = 0;
+int64_t WorkGroupDriverQueue::_ideal_runtime_ns(workgroup::WorkGroupDriverSchedEntity* wg_entity) const {
+    return SCHEDULE_PERIOD_PER_WG_NS * _ready_wg_entities.size() * wg_entity->cpu_limit() / _sum_cpu_limit;
+}
+
+void WorkGroupDriverQueue::_update_bandwidth_period() {
```

Comment:
`schedule_period` ?

Reply:
Bandwidth control is from linux kernel hard limit for cgroup.
Let me change it to `bandiwdth_control_period`